### PR TITLE
perf(ext/http): optimize auto cleanup of request resource

### LINKED ISSUE #11963: Performance regression after fixing resource leak in HTTP bindings.
#11955 caused a performance regression in HTTP bindings.

image0

### PR DESCRIPTION
**Pull request by piscisaureus**
Fixes #11963.

### CODE REVIEW

File: core/ops_builtin.rs [hunk 1]
```diff
@@ -44,10 +45,24 @@ pub fn op_close(
...
+/// Try to remove a resource from the resource table. If there is no resource
+/// with the specified `rid`, this is a no-op.
+pub fn op_try_close(
+  state: &mut OpState,
+  rid: Option<ResourceId>,
+  _: (),
+) -> Result<(), AnyError> {
+  // TODO(@AaronO): drop Option after improving type-strictness balance in
+  // serde_v8.
+  let rid = rid.ok_or_else(|| type_error("missing or invalid `rid`"))?;
```
Comment (AaronO) [inline core/ops_builtin.rs:64]: I would ignore this error too

**Review by AaronO — DISMISSED** [review]
Pretty much what I was planning to do (on Monday ...), looks good, though I would also allow rid to be null/undefined too so it's really fire and forget "close if there is something to close and be silent"

File: core/ops_builtin.rs (hunk 1 above)
Comment (bartlomieju) [reply core/ops_builtin.rs:64]: What's the point of calling into Rust if `rid` is `null` or `undefined`? Seems like overly relaxed API

Comment (piscisaureus) [reply core/ops_builtin.rs:64]: > What's the point of calling into Rust if `rid` is `null` or `undefined`? Seems like overly relaxed API

I don't quite follow. Did you mean "overly strict"?

I agree that we shouldn't call into rust when `rid` is null or undefined, so in order to detect those cases we *should* throw when the user does that.

Comment (AaronO) [reply core/ops_builtin.rs:64]: Agreed, but IIRC there are some tests where we call `Deno.close()` and other things of the sort on potentially nullish values, so when rewriting the op-layer I remained compatible the prior approach. We could add that check in `core.js`, will revisit that and other op-layer changes in the coming weeks

Comment (bartlomieju) [reply core/ops_builtin.rs:64]: > > What's the point of calling into Rust if `rid` is `null` or `undefined`? Seems like overly relaxed API
> 
> I don't quite follow. Did you mean "overly strict"?
> 
> I agree that we shouldn't call into rust when `rid` is null or undefined, so in order to detect those cases we _should_ throw when the user does that.

I meant I'm not in favor of the proposed changed and the API is good as it is now.

> Agreed, but IIRC there are some tests where we call Deno.close() and other things of the sort on potentially nullish values, so when rewriting the op-layer I remained compatible with the previous API. We could add that check in core.js, will revisit that and other op-layer changes in the coming weeks

But this is a new API and it doesn't interfere with `Deno.close()`, IMO the check should be done in JS for nullish values before calling, it's surely faster than crossing the boundary between V8 and Rust.

Comment (bartlomieju) [reply core/ops_builtin.rs:64]: That said, I'm not entirely sure why the type of argument to these ops is `Option<ResourceId>`... If it was `ResourceId` then nullish values would be handled by `serde_v8`, wouldn't they?

**Comment by piscisaureus** [comment]
```
  "req_per_sec": {
    "deno_tcp": 63747,
    "node_proxy": 2487,
    "node_http": 20239,
    "core_http_json_ops": 75343,
    "deno_proxy_tcp": 24360,
    "node_tcp": 54886,
    "node_proxy_tcp": 25079,
    "deno_http": 26781,
    "deno_http_native": 26950,
    "hyper": 70560
  },
```

File: core/ops_builtin.rs (hunk 1 above)
Comment (piscisaureus) [reply core/ops_builtin.rs:64]: > If it was `ResourceId` then nullish values would be handled by `serde_v8`, wouldn't they?

Good point - I just copypasta'd from `op_close()`. @AaronO?

**Review by bartlomieju — APPROVED** [review]
LGTM

File: core/ops_builtin.rs (hunk 1 above)
Comment (piscisaureus) [reply core/ops_builtin.rs:64]: > If it was `ResourceId` then nullish values would be handled by `serde_v8`, wouldn't they?

It seems that serde_v8 coerces to `0` in that case. With the proposed change:

```
piscisaureus@dumbo:~/d/deno 🦕 ./target/debug/deno
Deno 1.13.2
exit using ctrl+d or close()
> Deno.core.resources()
{ "0": "stdin", "1": "stdout", "2": "stderr" }
> Deno.core.tryClose()
undefined
> Deno.core.resources()
{ "1": "stdout", "2": "stderr" }
> 
```

Comment (AaronO) [reply core/ops_builtin.rs:64]: Yeah `serde_v8` deserializes ints via `.integer_value()` / `v8::Value::IntegerValue`, hence the note about type-strictness & coercion. For example other non `true` values coerce to `false` when deserializing a `bool`. Parts of our ops depended on that implicit coercion, so I added it to match existing behaviour vs changing all those call sites and delaying the `serde_v8` op-layer work.

I'll revisit this consistency and other op-layer work in the coming weeks

Comment (bartlomieju) [reply core/ops_builtin.rs:64]: @AaronO sounds good, thanks!

**Pull request merged**